docs/man/dbl2png.1
.TH dbl2png 1 "25 Nov 2009" "" "Ming utils"
.\" $Id$
.SH NAME
dbl2png - convert a dbl file to an png file
.SH SYNOPSIS
.B dbl2png [--verbose] <file.dbl>
.SH DESCRIPTION
.B dbl2png
is a command line tool for converting a dbl file into a png file. Output is written to file.png.
H OPTIONS
.TP
\fB\--verbose\fR
displays additional information during processing.
.TP
\fB<file.dbl>\fR
Name of dbl format file to use as input.
.SH AUTHOR
Written by Albrecht Kleine <kleine@ak.sax.de>
.SH COPYRIGHT
Code is released under the GNU GPL license.


NEWS
0.5.0 - ??

	* fixed SWFMatrix scaling if scale = 0.0
	* added delete and delete2 support in asm blocks
	* a lot of updates and improvements in the Perl interface
	* a lot of documentation (pod) was added to Perl files
	* Support for manually specifying MP3 delaySeek for SWFSound
	  and SWFSoundStream

0.4.0.rc2 - 2008-09-17

	* SWFVideoStream allow manual frame control
	* Autoconf perl version
	* MSVC compile fixes
	* removed src/test.c
	* Embedded video in movie clips
	* Seek in Embedded video streams
	* Gradient / Fill fixes
	* Sound / Soundstream fixes

0.4.0.rc1 - 2008-07-05

	* Perl extension updates 
	* Python extension updates
	* Do not force bgColor
	* Redirect SWF_warn and SWF_error to stderr
	* Propper checks for YACC and LEX
	* Complete Bitmap fillstyles
	* Font fixes
	* Actioncompiler fixes (delete, built-in functions, callFrame -> call())
	* Fixed tag order issues
	* Ming defaults now to local-with-filesystem sandbox for SWF>=8
	* Add support for init actions in makeswf
	* SWF8 / 9 support for fromswf / prebuiltclip
	* implements ASM op
	* ASFunction()
	* AS Class support
	* Removed SWFFontCharacter destructor from PHP wrapper
	* py_ext update
	* Class members
	* outputBlock() usese length and offset from SWF_Parserstruct
	* SWFDisplayItem_flush(): immediately write display items.
	* Removed ttftofft
	* Improved ButtonRecord handling
	* SWFSound_fromSoundStream()
	* Many improvements in decompiler and swftoXXX tools
	* C++ might throw a SWFException
	* Support for embedded Nellymoser sound
	* AS: in with-blocks legacy functions o not work
	* Compiles again with MSVC
	* simplified SWFSound usage with mp3 files
	* SWFSoundStream_getDuration replaces SWFSoundStream_getFrames()
	* perl_ext build changes (build out of tree)
	* API float->double transition
	* php_ext: SWFInput can wrap raw byte-streams 
	* php_ext: API: external data can be passed as file-resource, SWFInput or filename
	* removed newSWFFont() as public function (causes various crashes)
	* removed new SWFFont(...) usage with browserfonts in script languages. Use SWFBrowserFont class.

0.4.0.beta5 - 2007-09-18

	* Add support for targetPath asm opcode.
	* Flash 8: filter and blend modes for buttons
	* Flash 8: SWF_DEFINEMORPHSHAPE2
	* Flash 8: SWFShape rendering hints
	* Flash 8: new gradient features
	* SWFFillStyle is reusable now and calculates propper GradientMatrix
	* setLeft/RightFillStyle adds fill to the shape if necessary
	* Handling fonts as shapes
	* Basic ttf font support 
	* Typesafe SWFMovie_add / replace
	* Add Envelopes to SoundInstance
	* ABC (AS3) parser and output
	* Parse SWF_SERIALNUMBER
	* Parse SWF9 tags
	* SWF9: DefineBinaryData
	* SWF9: SymbolClass
	* SWF9: DefineSceneData

0.4.0.beta4 - 2007-03-27

	* Fix handling of the special 'undefined' value.
	* Fix 'delete' syntax in action compiler.
	* makeswf now handles bitmaps.
	* Many improvements in decompiler.
	* Build scripts cleanup (including check rule).
	* Add support for 'extends' and proper 'push' asm opcodes in
	  ActionScript compiler.
	* Generate DefineFunction2 when target version > 6.
	* Use GOTOFRAME, GOTOFRAME2 and GOTOLABEL when using gotoAndPlay()
	  or gotoAndStop() w/out an object.
	* Video updates: added Flash8 codecs, generic flv parser
	* Soundstream: accept flv audio streams (mp3 only for now)
	* Flash 8: PlaceObject3
	* Flash 8: Filter and Blendmode
	* Write proper Flash 8 files
	* Added movie control tags: FileAttributes, Metadata, 
	  ScriptLimits, ImportAssets2, SetTabIndex
	* Flash 8: ScalingGrid
	* Parser / listswf updates
	* PHP / Perl bindings update
	* Flash 8: DefineShape4 / Linestyle2
	* SWFAction cleanup and API cleanup
	* SWF_warnOnce()
	* SWFInitAction

0.4.0.beta3 - 2006-12-10

	* Fix version to not use dashes to make rpmbuild happy
	* autotool tcl extension
	* coalesce multiple ming.i files into a single file at the top
	* swig output updated for python & tcl
	* Misc decompiler fixes
	* makeswf now handles precompiled clips
	* Make sure the extension source show up in the dist tarballs
	* compiler fix for asm {} blocks

0.4.0-beta2 - 2006-11-21

	* Preliminary support for building RPM packages
	* New PHP code that builds for PHP$ and PHP5
	* Added missing SWFMovieClip class in PHP
	* New tool, ttftofft, for producing ming formast fonts
	* Allow output of SWF version 8 
	* Better SWF parser (utilities).
	* NEW function: SWFMovie_protectWithPassword()
	* preprocessor products of 'makeswf' are now named after
	  output filename and frame# (ie: "out.swf.frame0.pp")
	* Automake-based build, for portability's sake (libtool)
	* Ming_setWarnFunction and Ming_setErrorFunction now return
	  the previously-set value.
	* More documentation !
	* Start of a test suite!
	* makeswf now provides an __SWF_VERSION__ macro for use
	  in cross-version ActionScript code.
	* NEW functions: SWFDisplayItem_getMatrix(),
	  SWFPosition_getMatrix(), SWFMatrix_getScaleX(),
	  SWFMatrix_getRotate0(), SWFMatrix_getRotate1(),
	  SWFMatrix_getScaleY(), SWFMatrix_getTranslateX(),
	  SWFMatrix_getTranslateY(), SWFShape_moveScaledPenTo().

0.3.1 - ?

	* Fixed libpng-less builds
	* Fixed default error handler to exit with a FAILURE code

0.3.0 - 2006-02-15

	* Fixed bug in empty case: statements
	* Updated ch language bindings

0.3beta2 - 2006-01-30

        * C# support
        * Garbage collection support
        * C++ / PHP / Java extensions fixes
        * Autotools / configure support
	* New SWFPrebuiltClip class - include external SWFs as MovieClips
	* Command line actionscript compiler
	* Decompiler redesign
        * Fixes and cleanups all over the place
	* Case-sensitive actionscript compiler
	* Language bindings are now split out into seperate tarballs
	* New utils based on common code and new parsing framework

0.3beta1 - 2004-02-02

0.2a - ???


util/dbl2png.c
#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>

#include <ming_config.h>

#include <png.h>

#include <zlib.h>

int verbose = 0;
typedef unsigned char byte;

void error(char *s, ...)
{
  va_list ap;
  va_start(ap, s);
  vprintf(s, ap);
  va_end(ap);
  putchar('\n');
  exit(1);
}

void usage()
{
  printf("dbl2png - convert a dbl file to an png file\n");
  printf("\nUsage: dbl2png [--verbose] <file.dbl>\n");
  printf("where file.dbl is your dbl file.  Writes to file.png.\n");
  exit(1);
}

struct pngdata
{
  png_uint_32 width;
  png_uint_32 height;
  int bit_depth;
  int color_type;
  int num_palette;
  int channels;
  png_colorp palette;
  unsigned char *data;
};


struct pngdata readDBL(FILE *fp)
{
  struct pngdata png;
  byte header[13],*data;
  unsigned long outsize,readsize=0,datsize;
  int i, ret;

  if(!fp)
    error("Couldn't open file!\n");
  
  ret = fread(header, 1, 13, fp);
  if(ret != 13)
    error("Couldn't read header\n");

  if (header[3]==2)
  {
   if(verbose) 
    printf("image data RGBA\n");
   png.color_type=PNG_COLOR_TYPE_RGB_ALPHA;
  }
  else
  {
   if (header[3]==1)
   {
    if(verbose) 
     printf("image data RGB\n");
    png.color_type=PNG_COLOR_TYPE_RGB;
   }
   else
    error("unexpected 3rd byte in header");
  }
  outsize=(header[4]<<24)+(header[5]<<16)+(header[6]<<8)+header[7];
  if(verbose)
    printf("outsize=%ld \n",outsize);

  png.bit_depth=8; 					// Can't deal with bit depth !=8
  png.width= header[9] +(header[10]<<8);
  png.height=header[11]+(header[12]<<8);

  if (header[8]==3)
  {
    png.color_type=PNG_COLOR_TYPE_PALETTE;
    readsize=outsize-6 ;
    png.num_palette=0;
    i=  fread(&png.num_palette,  1,  1 ,fp); 
    if (verbose) 
    {
      printf("image has PALETTE\npalette size=%d\n",png.num_palette+1);
      printf("size %d x %d\n", (int)png.width, (int)png.height);
    }
  }
  else
  {
   if (header[8]!=5)
    error("unexpected 8th byte in header");
   else
   {
    readsize=outsize-5 ;
    if(verbose)
    {
     printf("size %d x %d\n", (int)png.width, (int)png.height);
    }
   }
  }
  if (!readsize)
   error("internal error");
  //
  // ------ now we go --------
  //
  data=malloc( readsize );
  if (!data)
  {
   error("memory allocation");
  }
  datsize=fread(data, 1, readsize, fp);
  fclose(fp);
  if (datsize != readsize)
  {
    free(data);
    error("Didn't read all of the file! %d  vs %d",datsize , readsize);
  }
  if (png.color_type==PNG_COLOR_TYPE_PALETTE)			// for palette bit_depth==8
   datsize = (png.num_palette+1) * 3 + png.width * png.height;	// palette of RGB style
  else
   datsize = png.width *  png.height * 4;			// RGBA
  png.data=malloc(datsize);
  if (!png.data)
  {
   free(data);
   error("memory allocation");
  }

  /* uncompress the data */
  i=uncompress (png.data, &datsize, data, readsize );
  if (i)
  {
   printf("uncompress call result = %d\n",i);
   free(data);
   error("uncompress error");
  }
  if(verbose)
  {
    printf("unpacked data size t=%ld byte\n",datsize);
  }
  free(data);
  return png;
}

// callback function:
// reverse apply alpha, bytes shifting (see png2dbl.c for details)
// expecting pixel_depth==32, bit_depth==8, color_type==PNG_COLOR_TYPE_RGB_ALPHA
void alpha_apply(png_structp png_ptr, png_row_infop row_info, png_bytep p)
{
  int i;
  byte a, r, g, b;
  for(i=0;i<row_info->width;i++)
  {
   a = p[0];
   r = p[1];
   g = p[2];
   b = p[3];
   *p++ = (r<<8)/(a+1);
   *p++ = (g<<8)/(a+1);
   *p++ = (b<<8)/(a+1);
   *p++ = a;
  }
}


void writePNG(FILE *fp, struct pngdata png)
{
  png_structp png_ptr;
  png_infop info_ptr;
  int i,chan;
  byte *ptr=png.data; 

  if(!fp)
    error("Couldn't open file!\n");

  png_ptr = png_create_write_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
  if(!png_ptr)
    error("Couldn't create png_ptr\n");

  info_ptr = png_create_info_struct(png_ptr);
  if(!info_ptr)
  {
    png_destroy_write_struct(&png_ptr,(png_infopp)NULL);
    error("Couldn't create info_ptr\n");
  }

  if (setjmp(png_jmpbuf(png_ptr)))
  {
    png_destroy_write_struct(&png_ptr, &info_ptr);
    fclose(fp);
    error("Mystery error (longjmp called)\n");
  }

  png_init_io(png_ptr, fp);
  png_set_IHDR(png_ptr, info_ptr, png.width, png.height,
  		png.bit_depth, png.color_type, PNG_INTERLACE_NONE, 
  		PNG_COMPRESSION_TYPE_DEFAULT, PNG_FILTER_TYPE_DEFAULT);
  if (png.color_type==PNG_COLOR_TYPE_PALETTE)
    png_set_PLTE(png_ptr, info_ptr,(png_colorp)png.data, png.num_palette+1);
  png_write_info(png_ptr, info_ptr);

  //
  // ------ now we go --------
  //
  chan=png_get_channels(png_ptr, info_ptr);
  if (verbose)  
    printf("channel count=%d\n",chan);
  
  if(png.color_type == PNG_COLOR_TYPE_RGB_ALPHA )
  {
   png_set_write_user_transform_fn(png_ptr, alpha_apply);
   for (i=0;i<png.height;i++)
   { 
      png_write_row(png_ptr,ptr);
      ptr+=png.width * chan;				// 4
   }
  }
  if(png.color_type == PNG_COLOR_TYPE_RGB)
  {
   png_set_filler(png_ptr, 0, PNG_FILLER_BEFORE);
   for (i=0;i<png.height;i++)
   { 
      png_write_row(png_ptr,ptr);
      ptr+=png.width * 4;
   }
  }
  if(png.color_type == PNG_COLOR_TYPE_PALETTE)
  {
    ptr+=(png.num_palette+1) * 3; 			// for palette bit_depth==8 of RGB style
    for (i=0;i<png.height;i++)
    {
      png_write_row(png_ptr,ptr);
      ptr+=png.width * chan;				// 1
    }   
  }

  png_write_end(png_ptr,NULL);
  png_destroy_write_struct(&png_ptr, &info_ptr);
  free(png.data);
  fclose(fp);
}


int main(int argc, char *argv[])
{
  int len;
  char *outfile;
  struct pngdata png;

  if (argc > 2 && strcmp(argv[1], "--verbose")==0) {
	verbose = 1;
	argc--;
	argv++;
  }

  if(argc < 2)
    usage();

  len = strlen(argv[1]);

  if(strcmp(argv[1]+len-4, ".dbl") != 0)
    usage();

  if(argc < 3)
  {
    outfile = strdup(argv[1]);
    outfile[len-3] = 'p';
    outfile[len-2] = 'n';
    outfile[len-1] = 'g';
  }
  else
    outfile = argv[2];

  png = readDBL(fopen(argv[1], "rb"));
  writePNG(fopen(outfile, "wb"), png);
  exit(0);
}
